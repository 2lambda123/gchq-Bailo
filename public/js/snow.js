const TOTAL_SNOWFLAKES = 200

let snowDiv = document.getElementById('snowContainer')

function embRand(a, b) {
  return Math.floor(Math.random() * (b - a + 1)) + a
}

if (!snowDiv) {
  const snowflakes = new Array(TOTAL_SNOWFLAKES).fill(`<i class="snow"></i>`).join('\n')

  const snow = `
    .snow {
      position: absolute;
      width: 10px;
      height: 10px;
      background: white;
      border-radius: 50%;
      margin-top:-10px;
    }

    #snowContainer {
      position: fixed;
      left: 0;
      top: 0;
      bottom: 0;
      width: 100vw;
      height: 100vh;
      overflow: hidden;
      z-index: 9999999;
      pointer-events: none;
    }
  `

  const transitions = new Array(TOTAL_SNOWFLAKES)
    .fill(undefined)
    .map((_, i) => {
      const rndX = embRand(0, 1000000) * 0.0001
      const rndO = embRand(-100000, 100000) * 0.0001
      const rndT = (embRand(3, 8) * 10).toFixed(2)
      const rndS = (embRand(0, 10000) * 0.0001).toFixed(2)

      return `
      .snow:nth-child(${i}) {
        opacity: ${(embRand(1, 10000) * 0.0001).toFixed(2)};
        transform: translate(${rndX.toFixed(2)}vw,-10px) scale(${rndS});
        animation: fall-${i} ${embRand(10, 30)}s -${embRand(0, 30)}s linear infinite;
      }

      @keyframes fall-${i}{
        ${rndT}%{
          transform: translate(${(rndX + rndO).toFixed(2)}vw,${rndT}vh) scale(${rndS});
        } to {
          transform: translate(${(rndX + rndO / 2).toFixed(2)}vw, 105vh) scale(${rndS});
        }
      }
    `
    })
    .join('\n')

  snowDiv = document.createElement('div')
  snowDiv.id = 'snowContainer'

  snowDiv.innerHTML = `
    <style>
      ${snow}
      ${transitions}
    </style>
    ${snowflakes}
  `
  document.body.appendChild(snowDiv)
}
