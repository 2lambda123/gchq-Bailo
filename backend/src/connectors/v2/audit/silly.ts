import { Request } from 'express'

import { AccessRequestDoc } from '../../../models/v2/AccessRequest.js'
import { FileInterface, FileInterfaceDoc } from '../../../models/v2/File.js'
import { ModelCardInterface, ModelDoc, ModelInterface } from '../../../models/v2/Model.js'
import { ReleaseDoc } from '../../../models/v2/Release.js'
import { ReviewInterface } from '../../../models/v2/Review.js'
import { SchemaInterface } from '../../../models/v2/Schema.js'
import { ModelSearchResult } from '../../../routes/v2/model/getModelsSearch.js'
import log from '../../../services/v2/log.js'
import { BailoError } from '../../../types/v2/error.js'
import { AuditInfo, AuditInfoKeys, BaseAuditConnector } from './Base.js'

interface Outcome {
  Success: boolean
  Description: string
  Permitted?: boolean
}

export class SillyAuditConnector extends BaseAuditConnector {
  constructor() {
    super()
  }

  /**
   * Model Events
   */

  onCreateModel(req: Request, model: ModelDoc) {
    this.checkEventType(AuditInfo.CreateModel, req)
    const event = this.generateEvent(req, { id: model.id })
    log.info(event, 'Logging Event')
  }

  onViewModel(req: Request, model: ModelDoc) {
    this.checkEventType(AuditInfo.ViewModel, req)
    const event = this.generateEvent(req, { name: model.name })
    log.info(event, 'Logging Event')
  }

  onUpdateModel(req: Request, model: ModelDoc) {
    this.checkEventType(AuditInfo.UpdateModel, req)
    const event = this.generateEvent(req, { id: model.id })
    log.info(event, 'Logging Event')
  }

  onSearchModel(req: Request, models: ModelSearchResult[]) {
    this.checkEventType(AuditInfo.SearchModels, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: models.map((model) => model.id),
    })

    log.info(event, 'Logging Event')
  }

  onCreateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface) {
    this.checkEventType(AuditInfo.CreateModelCard, req)
    const event = this.generateEvent(req, { modelId, version: modelCard.version })
    log.info(event, 'Logging Event')
  }

  onViewModelCard(req: Request, modelId: string, modelCard: ModelCardInterface) {
    this.checkEventType(AuditInfo.ViewModelCard, req)
    const event = this.generateEvent(req, { modelId, version: modelCard.version })
    log.info(event, 'Logging Event')
  }

  onUpdateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface) {
    this.checkEventType(AuditInfo.UpdateModelCard, req)
    const event = this.generateEvent(req, { modelId, version: modelCard.version })
    log.info(event, 'Logging Event')
  }

  onViewModelCardRevisions(req: Request, modelCards: ModelCardInterface[]) {
    this.checkEventType(AuditInfo.ViewModelCardRevisions, req)
    const event = this.generateEvent(req, { url: req.originalUrl, results: modelCards.map((model) => model.version) })
    log.info(event, 'Logging Event')
  }

  onCreateFile(req: Request, file: FileInterfaceDoc) {
    this.checkEventType(AuditInfo.CreateFile, req)
    const event = this.generateEvent(req, { id: file._id, modelId: file.modelId })
    log.info(event, 'Logging Event')
  }

  onViewFiles(req: Request, modelId: string, files: FileInterface[]) {
    this.checkEventType(AuditInfo.ViewFiles, req)
    const event = this.generateEvent(req, { modelId, results: files.map((file) => file.name) })
    log.info(event, 'Logging Event')
  }

  onDeleteFile(req: Request, modelId: string, fileId: string) {
    this.checkEventType(AuditInfo.DeleteFile, req)
    const event = this.generateEvent(req, { modelId, fileId })
    log.info(event, 'Logging Event')
  }

  /**
   * Release Events
   */

  onCreateRelease(req: Request, release: ReleaseDoc) {
    this.checkEventType(AuditInfo.CreateRelease, req)
    const event = this.generateEvent(req, { modelId: release.modelId, semver: release.semver })
    log.info(event, 'Logging Event')
  }

  onViewRelease(req: Request, release: ReleaseDoc) {
    this.checkEventType(AuditInfo.ViewRelease, req)
    const event = this.generateEvent(req, { modelId: release.modelId, semver: release.semver })
    log.info(event, 'Logging Event')
  }

  onUpdateRelease(req: Request, release: ReleaseDoc) {
    this.checkEventType(AuditInfo.UpdateRelease, req)
    const event = this.generateEvent(req, { modelId: release.modelId, semver: release.semver })
    log.info(event, 'Logging Event')
  }

  onDeleteRelease(req: Request, modelId: string, semver: string) {
    this.checkEventType(AuditInfo.DeleteRelease, req)
    const event = this.generateEvent(req, { modelId, semver })
    log.info(event, 'Logging Event')
  }

  onSearchReleases(req: Request, releases: ReleaseDoc[]) {
    this.checkEventType(AuditInfo.SearchReleases, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: releases.map((release) => ({ modelId: release.modelId, semver: release.semver })),
    })
    log.info(event, 'Logging Event')
  }

  onCreateAccessRequest(req: Request, accessRequest: AccessRequestDoc) {
    this.checkEventType(AuditInfo.CreateAccessRequest, req)
    const event = this.generateEvent(req, { id: accessRequest.id })
    log.info(event, 'Logging Event')
  }

  onViewAccessRequest(req: Request, accessRequest: AccessRequestDoc) {
    this.checkEventType(AuditInfo.ViewAccessRequest, req)
    const event = this.generateEvent(req, { id: accessRequest.id })
    log.info(event, 'Logging Event')
  }

  onUpdateAccessRequest(req: Request, accessRequest: AccessRequestDoc) {
    this.checkEventType(AuditInfo.UpdateAccessRequest, req)
    const event = this.generateEvent(req, { id: accessRequest.id })
    log.info(event, 'Logging Event')
  }

  onDeleteAccessRequest(req: Request, accessRequestId: string) {
    this.checkEventType(AuditInfo.DeleteAccessRequest, req)
    const event = this.generateEvent(req, { accessRequestId })
    log.info(event, 'Logging Event')
  }

  onSearchAccessRequests(req: Request, accessRequests: AccessRequestDoc[]) {
    this.checkEventType(AuditInfo.SearchAccessRequests, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: accessRequests.map((accessRequest) => ({
        id: accessRequest.id,
      })),
    })
    log.info(event, 'Logging Event')
  }

  onSearchReviews(req: Request, reviews: (ReviewInterface & { model: ModelInterface })[]) {
    this.checkEventType(AuditInfo.SearchReviews, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: reviews.map((review) => ({
        modelId: review.modelId,
        ...(review.semver && { semver: review.semver }),
        ...(review.accessRequestId && { semver: review.accessRequestId }),
      })),
    })
    log.info(event, 'Logging Event')
  }

  onCreateReviewResponse(req: Request, review: ReviewInterface) {
    this.checkEventType(AuditInfo.CreateAccessRequest, req)
    const event = this.generateEvent(req, {
      modelId: review.modelId,
      ...(review.semver && { semver: review.semver }),
      ...(review.accessRequestId && { semver: review.accessRequestId }),
    })
    log.info(event, 'Logging Event')
  }

  onError(req: Request, error: BailoError) {
    const outcome =
      error.code === 403
        ? {
            Description: 'User does not have permission to execute the request',
            Permitted: false,
            Success: false,
          }
        : {
            Description: error.message,
            Success: false,
          }
    const event = this.generateEvent(req, { url: req.originalUrl, httpMethod: req.method }, outcome)

    log.info(event, 'Logging Event')
  }

  generateEvent(req: Request, resourceInfo: object, outcome?: Outcome) {
    return {
      EventTime: new Date().toString(),
      EventSource: {
        System: 'BAILO',
        Device: {
          Hostname: 'http://localhost:8080',
        },
        Client: req.ip,
        User: req.user,
      },
      EventDetail: {
        typeId: req.audit.typeId,
        description: req.audit.description,
        resource: JSON.stringify(resourceInfo),
        ...(outcome && { outcome }),
      },
    }
  }

  onCreateSchema(req: Request, schema: SchemaInterface) {
    this.checkEventType(AuditInfo.CreateSchema, req)
    const event = this.generateEvent(req, { id: schema.id })
    log.info(event, 'Logging Event')
  }

  onViewSchema(req: Request, schema: SchemaInterface) {
    this.checkEventType(AuditInfo.ViewSchema, req)
    const event = this.generateEvent(req, { id: schema.id })
    log.info(event, 'Logging Event')
  }

  onSearchSchemas(req: Request, schemas: SchemaInterface[]) {
    this.checkEventType(AuditInfo.SearchSchemas, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: schemas.map((schema) => ({ id: schema.id })),
    })
    log.info(event, 'Logging Event')
  }

  onSearchImages(req: Request, images: { repository: string; name: string; tags: string[] }[]) {
    this.checkEventType(AuditInfo.SearchImages, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: images.map((image) => ({ repository: image.repository, name: image.name })),
    })
    log.info(event, 'Logging Event')
  }

  checkEventType(auditInfo: AuditInfoKeys, req: Request) {
    if (auditInfo.typeId !== req.audit.typeId && auditInfo.description !== req.audit.description) {
      throw new Error(`Audit: Expected type '${JSON.stringify(auditInfo)}' but recieved '${JSON.stringify(req.audit)}'`)
    }
  }
}
