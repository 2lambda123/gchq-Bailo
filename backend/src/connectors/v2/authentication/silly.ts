import { Request } from 'express'

import { UserDoc } from '../../../models/v2/User.js'
import { fromEntity, toEntity } from '../../../utils/v2/entity.js'
import { BaseAuthenticationConnector } from './Base.js'

const SillyEntityKind = {
  User: 'user',
  Group: 'group',
} as const

export class SillyAuthenticationConnector extends BaseAuthenticationConnector {
  constructor() {
    super()
  }

  async getUserFromReq(_req: Request) {
    return {
      dn: 'user',
    }
  }

  async isAdmin(_user: UserDoc) {
    return true
  }

  async queryEntities(_query: string) {
    return [
      {
        kind: SillyEntityKind.User,
        entities: [toEntity(SillyEntityKind.User, 'user1'), toEntity(SillyEntityKind.User, 'user2')],
      },
      {
        kind: SillyEntityKind.Group,
        entities: [toEntity(SillyEntityKind.Group, 'group1'), toEntity(SillyEntityKind.Group, 'group2')],
      },
    ]
  }

  async getEntities(user: UserDoc) {
    return [toEntity(SillyEntityKind.User, user.dn)]
  }

  async getUserInformation(entity: string): Promise<{ email: string }> {
    const { kind, value } = fromEntity(entity)

    if (kind !== SillyEntityKind.User) {
      throw new Error(`Cannot get user information for a non-user entity: ${entity}`)
    }

    return {
      email: `${value}@example.com`,
    }
  }

  async getEntityMembers(entity: string): Promise<string[]> {
    const { kind } = fromEntity(entity)
    switch (kind) {
      case SillyEntityKind.User:
        return [entity]
      case SillyEntityKind.Group:
        return [toEntity(SillyEntityKind.User, 'user1'), toEntity(SillyEntityKind.User, 'user2')]
      default:
        throw new Error(`Unable to get members, entity kind not recognised: ${entity}`)
    }
  }
}
