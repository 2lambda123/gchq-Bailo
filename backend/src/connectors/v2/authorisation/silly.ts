import { AccessRequestDoc } from '../../../models/v2/AccessRequest.js'
import { ModelDoc } from '../../../models/v2/Model.js'
import { ReleaseDoc } from '../../../models/v2/Release.js'
import { UserDoc } from '../../../models/v2/User.js'
import { fromEntity, toEntity } from '../../../utils/v2/entity.js'
import { AccessRequestActionKeys, BaseAuthorisationConnector, ModelActionKeys, ReleaseActionKeys } from './Base.js'

const SillyEntityKind = {
  User: 'user',
  Group: 'group',
} as const

export class SillyAuthorisationConnector extends BaseAuthorisationConnector {
  async userModelAction(_user: UserDoc, _model: ModelDoc, _action: ModelActionKeys) {
    // With silly authorisation, every user can complete every action.
    return true
  }

  async userReleaseAction(
    _user: UserDoc,
    _model: ModelDoc,
    _release: ReleaseDoc,
    _action: ReleaseActionKeys,
  ): Promise<boolean> {
    // With silly authorisation, every user can complete every action.
    return true
  }

  async userAccessRequestAction(
    _user: UserDoc,
    _model: ModelDoc,
    _accessRequest: AccessRequestDoc,
    _action: AccessRequestActionKeys,
  ) {
    // With silly authorisation, every user can complete every action.
    return true
  }

  async getEntities(user: UserDoc) {
    return [toEntity(SillyEntityKind.User, user.dn)]
  }

  async getUserInformation(entity: string): Promise<{ email: string }> {
    const entityObject = fromEntity(entity)
    if (entityObject.kind !== SillyEntityKind.User) {
      throw new Error(`Cannot get user information for a non-user entity: ${entity}`)
    }
    return {
      email: `${entityObject.value}@example.com`,
    }
  }

  async getEntityMembers(entity: string): Promise<string[]> {
    if (fromEntity(entity).kind === SillyEntityKind.User) {
      return [entity]
    } else if (fromEntity(entity).kind === SillyEntityKind.Group) {
      return [toEntity(SillyEntityKind.User, 'user1'), toEntity(SillyEntityKind.User, 'user2')]
    } else {
      throw new Error(`Unable to get Entity Members. Entity kind not recognised: ${entity}`)
    }
  }
}
