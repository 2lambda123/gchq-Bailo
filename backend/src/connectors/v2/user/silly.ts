import { BaseUserConnector } from './index.js'

export class SillyUserConnector implements BaseUserConnector {
  constructor() {
    // do nothing
  }
}
