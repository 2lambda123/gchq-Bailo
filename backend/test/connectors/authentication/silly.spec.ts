import { Request } from 'express'
import { describe, expect, test } from 'vitest'

import { SillyAuthenticationConnector } from '../../../src/connectors/v2/authentication/silly.js'

describe('connectors > user > silly', () => {
  test('getUserFromReq', async () => {
    const connector = new SillyAuthenticationConnector()
    const user = await connector.getUserFromReq({} as Request)

    expect(user).toStrictEqual({ dn: 'user' })
  })

  test('queryEntities', async () => {
    const connector = new SillyAuthenticationConnector()
    const queryResult = await connector.queryEntities('abc')

    expect(queryResult).matchSnapshot()
  })
})
