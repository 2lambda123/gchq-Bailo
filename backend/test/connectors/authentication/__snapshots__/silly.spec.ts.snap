// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`connectors > user > silly > queryEntities 1`] = `
[
  {
    "entities": [
      "user:user",
      "user:user2",
    ],
    "kind": "user",
  },
  {
    "entities": [
      "group:group1",
      "group:group2",
    ],
    "kind": "group",
  },
]
`;
